src/driver/hermes_cdev.c

```diff
@@ -97,6 +104,80 @@ static int hermes_close(struct inode *inode, struct file *filp)
 	return 0;
 }
 
+static int hermes_fsync(struct file *filp, loff_t start, loff_t end, int datasync)
+{
+	struct hermes_env *env = filp->private_data;
+	struct hermes_dev *hermes = env->hermes;
+	struct hermes_cmd cmd = {
+		.req = {
+			.opcode = HERMES_OPCODE_RUN_PROG,
+			.cid = env->cid,
+			.run_prog.prog_slot = env->prog_slot,
+			.run_prog.data_slot = env->data_slot,
+		},
+	};
+	int eng = 0, res;
+
+	if (env->prog_slot < 0) {
+		dev_err(&hermes->dev,
+			"Program has not been downloaded to device. Aborting.\n");
+		return -EBADFD;
+	}
+
+	if (env->data_slot < 0) {
+		dev_err(&hermes->dev,
+			"No data has been transferred to device. Aborting.\n");
+		return -EBADFD;
+	}
+
+	pr_debug("opcode: 0x%x cid: 0x%x prog_slot: 0x%x data_slot: 0x%x\n",
+			cmd.req.opcode, cmd.req.cid, cmd.req.run_prog.prog_slot,
+			cmd.req.run_prog.data_slot);
+
+	memcpy_toio(&hermes->cmds[eng].req, &cmd.req, sizeof(cmd.req));
+	iowrite8(1, &hermes->cmds_ctrl[eng]);
+
```

Comment:
Certainly there must be something to wait for the program to complete here.... It won't be instant with the io write (even though it may be in qemu atm).

Reply:
Good point. Do you have any suggestion on how to do this?

Reply:
Probably an interrupt?

Reply:
The hardware should probably set a flag in a register when it is done, then trigger an interrupt.

The driver needs an interrupt handler and a (probably) a wait queue. When an interrupt happens it wakes up any thread waiting which would check the register for completion.